classes/observers/listeners/UsageEventLog.inc.php

```diff
@@ -0,0 +1,472 @@
+        $cachedInstitutionData[$hashedIp][$contextId] = $institutionIds;
+        $this->institutionDataCache->setEntireCache($cachedInstitutionData);
+        return $cachedInstitutionData[$hashedIp][$contextId];
+    }
+
+    /**
+    * Institution cache miss callback.
+    */
+    public function institutionDataCacheMiss(FileCache $cache): array
+    {
+        $cache->setEntireCache([]);
+        return [];
+    }
+
+    /**
+     * Retrieve Geo data (country, region, city) using IP and based on the site i.e. context settings
+     */
+    protected function getGeoData(Site $site, Context $context, string $ip, string $hashedIp, bool $flushCache): array
+    {
+        $enableGeoUsageStats = $site->getData('enableGeoUsageStats');
+        if (($enableGeoUsageStats != 'disabled') && ($context->getData('enableGeoUsageStats') !== null) && ($context->getData('enableGeoUsageStats') != $site->getData('enableGeoUsageStats'))) {
+            $enableGeoUsageStats = $context->getData('enableGeoUsageStats');
+        }
+
+        $country = $region = $city = null;
+        if ($enableGeoUsageStats != 'disabled') {
+            $geoIPArray = $this->getLocation($ip, $hashedIp, $flushCache);
+            $country = $geoIPArray['country'];
+            if ($enableGeoUsageStats == 'country+region+city' || $enableGeoUsageStats == 'country+region') {
+                $region = $geoIPArray['region'];
+                if ($enableGeoUsageStats == 'country+region+city') {
+                    $city = $geoIPArray['city'];
+                }
+            }
+        }
+        return [$country, $region, $city];
+    }
+
+    /**
+     * Get location based on the IP, use cache if exists.
+     *
+     * @param string $ip User IP
+     * @param string $hashedIp Hashed user IP
+     * @param bool $flush If true empty cache
+     *
+     * @return array Cached Geo data
+     *  [
+     *   hashedIP => [
+     *    'country' => string Country ISO code,
+     *    'region' => string Region ISO code
+     *    'city' => string City name
+     *   ]
+     *  ]
+     *
+     */
+    protected function getLocation(string $ip, string $hashedIp, bool $flush): array
+    {
+        $ip = '142.58.100.60';
+        if (!isset($this->geoDataCache)) {
+            $geoCacheManager = CacheManager::getManager();
+            $this->geoDataCache = $geoCacheManager->getCache('geoIP', 'all', [&$this, 'geoDataCacheMiss']);
+        }
+
+        if ($flush) {
+            // Salt and thus hashed IPs changed, empty the cache.
+            $this->geoDataCacheMiss($this->geoDataCache, 'ID');
+        }
+
+        $cachedGeoData = $this->geoDataCache->getContents();
+        if (array_key_exists($hashedIp, $cachedGeoData)) {
+            return $cachedGeoData[$hashedIp];
+        }
+
+        $reader = $countryIsoCode = $regionIsoCode = $cityName = null;
+        try {
+            $reader = new Reader($this->getGeoDBPath());
```

Comment:
If the `IPGeoDB.mmdb` file hasn't been downloaded by the `UpdateIPGeoDB` task before this runs, it'll result in...
```
PHP Fatal error:  Uncaught InvalidArgumentException: The file "/home/asmecher/git/ojs-main/files/usageStats/IPGeoDB.mmdb" does not exist or is not readable. in /home/asmecher/git/ojs-main/lib/pkp/lib/vendor/maxmind-db/reader/src/MaxMind/Db/Reader.php:83
Stack trace:
#0 /home/asmecher/git/ojs-main/lib/pkp/lib/vendor/geoip2/geoip2/src/Database/Reader.php(66): MaxMind\Db\Reader->__construct()
#1 /home/asmecher/git/ojs-main/lib/pkp/classes/observers/listeners/UsageEventLog.inc.php(395): GeoIp2\Database\Reader->__construct()
#2 /home/asmecher/git/ojs-main/lib/pkp/classes/observers/listeners/UsageEventLog.inc.php(346): PKP\observers\listeners\UsageEventLog->getLocation()
#3 /home/asmecher/git/ojs-main/lib/pkp/classes/observers/listeners/UsageEventLog.inc.php(146): PKP\observers\listeners\UsageEventLog->getGeoData()
#4 /home/asmecher/git/ojs-main/lib/pkp/classes/observers/listeners/UsageEventLog.inc.php(58): PKP\observers\listeners\UsageEventLog->prepareUsageEvent()
#5 /home/asmecher/git/ojs-main/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(441): PKP\observers\listeners\UsageEventLog->handle()
#6 /home/asmecher/git/ojs-main/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(249): Illuminate\Events\Dispatcher->Illuminate\Events\{closure}()
#7 /home/asmecher/git/ojs-main/lib/pkp/lib/vendor/laravel/framework/src/Illuminate/Foundation/helpers.php(450): Illuminate\Events\Dispatcher->dispatch()
```
The `catch` should be expanded to handle this kind of problem too.